### PR TITLE
Add CUDA async memory resource as an option

### PR DESCRIPTION
**Pull request by rongou**
This only adds an option to set the gpu memory pool to use RMM `cuda_async_memory_resource`. Additional work needs to be done to support UCX (e.g. https://github.com/openucx/ucx/pull/7204) and possibly GDS.

Depends on https://github.com/rapidsai/cudf/pull/9201.

Signed-off-by: Rong Ou <rong.ou@gmail.com>

### CODE REVIEW

**Review by abellina — APPROVED** [review]

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala [hunk 1]
```diff
@@ -411,12 +411,14 @@ object RapidsConf {
...
   val RMM_POOL = conf("spark.rapids.memory.gpu.pool")
-    .doc("Select the RMM pooling allocator to use. Valid values are \"DEFAULT\", \"ARENA\", and " +
-      "\"NONE\". With \"DEFAULT\", `rmm::mr::pool_memory_resource` is used; with \"ARENA\", " +
-      "`rmm::mr::arena_memory_resource` is used. If set to \"NONE\", pooling is disabled and RMM " +
-      "just passes through to CUDA memory allocation directly. Note: \"ARENA\" is the " +
-      "recommended pool allocator if CUDF is built with Per-Thread Default Stream (PTDS), " +
-      "as \"DEFAULT\" is known to be unstable (https://github.com/NVIDIA/spark-rapids/issues/1141)")
+    .doc("Select the RMM pooling allocator to use. Valid values are \"DEFAULT\", \"ARENA\", " +
+      "\"ASYNC\", and \"NONE\". With \"DEFAULT\", `rmm::mr::pool_memory_resource` is used; with " +
+      "\"ARENA\", `rmm::mr::arena_memory_resource` is used; with \"ASYNC\", " +
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala:416]: This doc is intended for end-users, and I think `rmm::mr::arena_memory_resource` is too much detail here.  I suggest simplifying this to say it uses the new CUDA stream-ordered memory allocator in CUDA 11.2+.

Comment (rongou) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala]: Done.

**Review by abellina — APPROVED** [review]

**Review by jlowe — APPROVED** [review]

**Comment by sameerz** [comment]
build

**Comment by rongou** [comment]
build

**Comment by rongou** [comment]
build

**Pull request merged**